### DISCUSSION TITLE
Discussion: Implement and apply an enumerated type

### DISCUSSION THREAD

**Discussion by yunkon-kim** [cloud-barista/cb-tumblebug#386] [General]
### [제안]
- Constants Variable 선언시 enum 형태로 사용하는 것을 어떨까요?
- 상수의 연관성을 나타낼 수 있어 개발 또는 리뷰할때 소스코드의 직관성을 높일 수 있을 것 같습니다.
- 관련 PR: #367(closed), #371(closed)

### [논의 포인트]
- 관련된 코드 전반에 대해 수정이 필요한데, 불필요한 공수 일 수 있습니다. (IDE에서 잘 처리해준다면 좋을 것 같습니다. ㅎㅎ)
- 향후에도 유용하게 쓰인다면 적용하고 아니라면 현상태 유지하는게 좋을 것 같습니다.
- 참고) 현재 Golang에서 enum을 지원하지 않기 때문에 약간의 트릭이 필요합니다.
  - [An idiomatic way to implement an enumerated type](https://yourbasic.org/golang/iota/)
  - [Golang에서 enum 지원이 필요하다는 제안](https://github.com/golang/go/issues/19814#issue-218638547)

#### [선언]
```
type ResourceType int

const (
    Image                ResourceType = iota
    SecurityGroup 
    Spec
)

func (rt ResourceType) String() string {
    return [...]string{"image", "securityGroup", "spec"}[rt]
}
```

#### [활용 예]
[Example on The Go Playground](https://play.golang.org/p/AXOTOYp7pTW)

**Comment by seokho-son** [comment]
좋은 의견 감사합니다!
단순히 상수로 정의해서 사용하는 방법은 다소 가독성이 떨어지긴 하겠네요.

Golang이 Enum을 직접적으로 제공하지 않는데, 나름대로 이유가 있을 것 같습니다. (예를 들어, `pkg를 활용하는 것으로 촉진하기 위함` `컴파일시 처리되는 것 방지` 이랄지..) 아직은 저도 정확하게는 파악이 안되네요.. ^^ 워낙 관련 디스커션이 길고 많아서.. ㅎㅎ

일단 가능한 Golang 네이티브한 방법을 찾아보고, 대안이 없으면 제안해주신 Enum 스타일의 방법을 적용해보면 어떨까요?

**Comment by yunkon-kim** [comment]
좋은 안 주셔서 감사합니다.

제가 제안한 Enum 스타일로 쓰는 방식은 보류하는 것이 좋을 것 같습니다.

그리고 아래의 계획을 다소 장기적으로 진행하고자 합니다.
1) Golang Native한 대안을 조금 더 탐색 또는 강구함
2) Golang에서 Enum을 어떤식으로 적용하는지 중간중간 확인
